draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -2715,108 +2715,127 @@ This approach increases the Hamming distance between inputs to different
+
+    To instantiate a suffix-only domain separated function Hso, compute
+
+        Hso(msg) = H(msg || DST_ext)
+
+    DST\_ext should be suffix-free encoded (e.g., by appending one byte
+    encoding the tag's length) to make it infeasible to find distinct
+    (msg, DST\_ext) pairs that hash to the same value.
+
+    This method ensures domain separation because all distinct invocations of
+    H have distinct suffixes.
+
+2.  (Prefix-suffix domain separation.)
+    This method can be used in the same cases as the suffix-only method.
+
+    To instantiate a prefix-suffix domain separated function Hps, compute
+
+        Hps(msg) = H(DST_ext || msg || I2OSP(0, 1))
+
+    DST\_ext should be prefix-free encoded (e.g., by prepending one byte
+    encoding the tag's length) to make it infeasible to find distinct
+    (msg, DST\_ext) pairs that hash to the same value.
+
+    This method ensures domain separation because
+    appending the byte I2OSP(0, 1) ensures that inputs to H inside Hps
+    are distinct from those inside expand\_message: the final byte of
+    DST\_prime encodes the length of DST, which is required to be nonzero
+    ({{domain-separation}}, requirement 2), and DST\_prime is always appended
+    to invocations of H inside expand\_message.
+
+3.  (Prefix-only domain separation.)
+    This method is only useful for domain separating invocations of H
+    from expand\_message\_xmd.
+    It does not give domain separation for expand\_message\_xof or for HMAC-H.
+
+    To instantiate a prefix-only domain separated function Hpo, compute
+
+        Hpo(msg) = H(DST_ext || msg)
+
+    In order for this method to give domain separation, DST\_ext should
+    be at least b bits long, where b is the number of bits output by the
+    hash function H.
+    In addition, at least one of the first b bits must be nonzero.
+    Finally, DST\_ext should be prefix-free encoded (e.g., by prepending
+    one byte encoding the tag's length) to make it infeasible to
+    find distinct (msg, DST\_ext) pairs that hash to the same value.
+
+    This method ensures domain separation as follows.
+    First, since DST\_ext contains at least one nonzero bit among its first b bits,
+    it is guaranteed to be distinct from the value Z\_pad
+    ({{hashtofield-expand-xmd}}, step 4), which ensures that all inputs to H
+    are distinct from the input used to generate b\_0 in expand\_message\_xmd.
+    Second, since DST\_ext is at least b bits long, it is almost certainly
+    distinct from the values b\_0 and strxor(b\_0, b\_(i - 1)), and therefore
+    all inputs to H are distinct from the inputs used to generate b\_i, i >= 1,
+    with high probability.
+
+4.  (XMD-HMAC domain separation.)
+    This method is useful for domain separating invocations of H inside
+    HMAC-H (i.e., HMAC {{RFC2104}} instantiated with hash function H) from
+    expand\_message\_xmd.
+    This is useful when HMAC-H is used with a non-secret key to
```

Comment:
Should we note something about HMAC-H when used with a secret key? (HPKE, for example, assumes HMAC is indifferentiable from a RO.)